packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -281,14 +282,25 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
           return;
         }
 
-        const maybeDoc = this.serializer.fromDbRemoteDocument(dbRemoteDoc);
-        if (!query.path.isPrefixOf(maybeDoc.key.path)) {
+        if (!query.path.isPrefixOf(new ResourcePath(key))) {
           control.done();
-        } else if (maybeDoc instanceof Document && query.matches(maybeDoc)) {
-          results = results.insert(maybeDoc.key, maybeDoc);
+          return;
         }
+
+        const decodeDoc = Promise.resolve().then(() => {
+          const maybeDoc = this.serializer.fromDbRemoteDocument(dbRemoteDoc);
+          if (maybeDoc instanceof Document && query.matches(maybeDoc)) {
+            results = results.insert(maybeDoc.key, maybeDoc);
+          }
+        });
+        promises.push(decodeDoc);
       })
-      .next(() => results);
+      .next(async () => {
+        await Promise.all(promises);
+      })
+      .next(() => {
+        return results;
+      });
```

Comment:
This could be condensed to:

```
.next(() => promise.all(promises))
.next(() => results);
```

or:
```
.next(async () => {
   await Promise.all(promises);
   return results;
});
```